### PR TITLE
s_server:  Report an error if init-connection fails without an attempt to read

### PR DESCRIPTION
**Pull request by faramir-dev**
…connection.

Fixes openssl#18047.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by bernd-edlinger** [comment]
note the typo in the commit message

**Comment by t8m** [comment]
There is one more way how you could possibly solve this issue. Set a callback function on the read bio with BIO_set_callback_ex() and if the callback is not called during the init connection call, it means the failure does not have anything with the connection attempt.

File: apps/s_server.c [hunk 1]
```diff
@@ -2328,6 +2328,30 @@ static void print_stats(BIO *bio, SSL_CTX *ssl_ctx)
...
+        if (p_counter != NULL)
+            ++*p_counter;
+        break;
+    default:
+        break;
+    }
+
+    if (s_debug) {
+        BIO_set_callback_arg(bio, (char* )bio_s_out);
+        bio_dump_callback(bio, cmd, argp, len, argi, argl, ret, processed);
```
Comment (t8m) [inline apps/s_server.c:2348]: Although the bio_dump_callback() does not modify the ret value we should do here `ret = (int)bio_dump....`

File: apps/s_server.c [hunk 2]
```diff
@@ -2328,6 +2328,30 @@ static void print_stats(BIO *bio, SSL_CTX *ssl_ctx)
...
+        BIO_set_callback_arg(bio, (char* )p_counter);
```
Comment (t8m) [inline apps/s_server.c:2349]: Should be `(char *)`

File: apps/s_server.c [hunk 3]
```diff
@@ -2328,6 +2328,30 @@ static void print_stats(BIO *bio, SSL_CTX *ssl_ctx)
                SSL_CTX_sess_get_cache_size(ssl_ctx));
 }
 
+static long count_reads_callback(BIO *bio, int cmd, const char *argp, size_t len,
+                                 int argi, long argl, int ret, size_t *processed)
+{
+    unsigned *p_counter = (unsigned*)BIO_get_callback_arg(bio);
```
Comment (t8m) [inline apps/s_server.c:2334]: `(unsigned *)`

File: apps/s_server.c [hunk 4]
```diff
@@ -2328,6 +2328,30 @@ static void print_stats(BIO *bio, SSL_CTX *ssl_ctx)
...
+    case BIO_CB_GETS:
+        if (p_counter != NULL)
+            ++*p_counter;
+        break;
+    default:
+        break;
+    }
+
+    if (s_debug) {
+        BIO_set_callback_arg(bio, (char* )bio_s_out);
```
Comment (mattcaswell) [inline apps/s_server.c:2347]: `(char *)`

File: apps/s_server.c (hunk 3 above)
Comment (mattcaswell) [reply apps/s_server.c:2334]: Please make this `unsigned int *` rather than just `unsigned *`

File: apps/s_server.c [hunk 5]
```diff
@@ -2737,8 +2758,24 @@ static int sv_body(int s, int stype, int prot, unsigned char *context)
              */
             if ((!async || !SSL_waiting_for_async(con))
                     && !SSL_is_init_finished(con)) {
+                /*
+                 * Count number of reads during init_ssl_connection.
+                 * It helps us to recognise configuration errors and errors
+                 * caused by a client.
+                 */
+                unsigned read_counter = 0;
```
Comment (mattcaswell) [inline apps/s_server.c:2766]: `unsigned int read_counter = 0;`

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Looks good. Some minor nits below

**Review by t8m — APPROVED** [review]

File: apps/s_server.c [hunk 6]
```diff
@@ -2737,8 +2758,24 @@ static int sv_body(int s, int stype, int prot, unsigned char *context)
              */
             if ((!async || !SSL_waiting_for_async(con))
                     && !SSL_is_init_finished(con)) {
+                /*
+                 * Count number of reads during init_ssl_connection.
+                 * It helps us to recognise configuration errors and errors
+                 * caused by a client.
+                 */
+                unsigned int read_counter = 0;
+                BIO_set_callback_arg(SSL_get_rbio(con), (char *)&read_counter);
```
Comment (paulidale) [inline apps/s_server.c:2767]: Blank line after declaration.

File: apps/s_server.c [hunk 7]
```diff
@@ -2737,8 +2758,24 @@ static int sv_body(int s, int stype, int prot, unsigned char *context)
              */
             if ((!async || !SSL_waiting_for_async(con))
                     && !SSL_is_init_finished(con)) {
+                /*
+                 * Count number of reads during init_ssl_connection.
+                 * It helps us to recognise configuration errors and errors
```
Comment (paulidale) [inline apps/s_server.c:2763]: What about _It helps us to distinguish configuration errors from errors caused by a client_?

File: apps/s_server.c [hunk 8]
```diff
@@ -2328,6 +2328,30 @@ static void print_stats(BIO *bio, SSL_CTX *ssl_ctx)
                SSL_CTX_sess_get_cache_size(ssl_ctx));
 }
 
+static long count_reads_callback(BIO *bio, int cmd, const char *argp, size_t len,
```
Comment (paulidale) [inline apps/s_server.c:2331]: Since we're being pedantic about the _int_ part: `long int`.

**Review by paulidale — APPROVED** [review]
Approved with the style nits fixed and possible with the changed comment (or a not changed comment).

**Comment by t8m** [comment]
@faramir-dev ping for fixups

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@paulidale still OK?

**Review by mattcaswell — APPROVED** [review]

**Pull request closed without merging**